machine.go

```diff
@@ -616,8 +616,36 @@ func (m *Machine) captureFifoToFile(logger *log.Entry, fifoPath string, w io.Wri
 			}
 		}()
 
+		rfds := &syscall.FdSet{}
+		// FD_SET taken from select.h of the linux kernel.
+		rfds.Bits[fd/strconv.IntSize] |= (1 << uint(fd%strconv.IntSize))
+
+		timeout := &syscall.Timeval{}
+		timeout.Usec = 10000000 // timeout in 10 microseconds
+		const maxRetries = 10
+		// Retry until there is data available on the file descriptor or that we've
+		// reached our retry limit.
+		for i := 0; i < maxRetries; i++ {
+			n, err := syscall.Select(fd+1, rfds, nil, nil, timeout)
```

Comment:
I'd prefer if we can find a different way to do this that doesn't involve `select`. The code written is already fairly complicated and it doesn't to my knowledge yet deal with issues mentioned in the `select(2)` docs like `EINTR` being returned if a signal interrupts the call, needing to reset the `Timeval` struct every time, or the mysterious phrase [`There may be other circumstances in which a file descriptor is spuriously reported as ready`](http://man7.org/linux/man-pages/man2/select.2.html#BUGS) (which implies that we could for unknown reasons still try to do `io.Copy` on a FIFO that hasn't been opened for writing yet).

In firecracker-containerd we just use the [`containerd/fifo`](https://github.com/containerd/fifo) package to open a FIFO and then later just call `io.Copy` on it without needing to deal with any sort of retries ourselves. Have you given that a try? It seems like there's potential it could simplify this code quite a bit.

Reply:
I don’t know if that package will work by looking at the code. I am pretty sure we will run into the same issue. I can try it, but I have a feeling it will be for naught

Reply:
Try opening it w/ just `syscall.O_RDWR` using `fifo.OpenFifo` on line 582. My understanding of the implementation in the `fifo` package is that when you do that, you won't block on the open call waiting for the other side (which is what we want) but then once you actually do the call to `io.Copy`, [the reader will internally wait for the other side to be opened or the object returned by `OpenFifo` to be closed (i.e. via line 599).](https://github.com/containerd/fifo/blob/bda0ff6ed73c67bfb5e62bc9c697f146b7fd7f13/fifo.go#L152-L162)

Reply:
@sipsma - Yea, it totally works! I was skeptical but didn't see that they implemented the `syscall.Conn` interface in raw which handles all the states for the fifo. Very nice implementation!

Reply:
Awesome! Yeah I'm very glad the containerd folks made that package...